File: src/__tests__/transport-request.test.ts
import fs from "node:fs";
import path from "node:path";
import { describe, expect, it } from "vitest";
import { verify } from "../shared/crypto.js";
import {
	signRosterRequest,
	signTransportRequest,
	TransportRequestSchema,
	transportRequestSigningBytes,
	verifyTransportRequest,
} from "../shared/enrollment.js";
import { assertCanonicalBytes } from "./_canonical-bytes.js";

////////////////////////////////
//  Transport request signing-bytes vectors
//
//  vectors.json is read by BOTH this suite and TransportRequestTest.kt (Kotlin), so the
//  hand-authored Kotlin twin (ProvisionOpsCrypto.transportRequestSigningBytes) cannot drift
//  from this TS source: the canonical bytes / signature either runtime derives differently
//  fails one of the two suites. The request is owner-signed proof of possession; evie resolves
//  the signer to a rooted owner and returns the gateway-bridge transport.

interface SignedVec<T> {
	value: T;
	signingBytes: string;
	signingBytesHex: string;
	signingBytesBase64: string;
	signature: string;
}

const vectors = JSON.parse(
	fs.readFileSync(path.join(__dirname, "../../tests/fixtures/transport-request/vectors.json"), "utf8"),
) as {
	ownerSignPub: string;
	ownerSignPriv: string;
	transport: SignedVec<{ signerSignPub: string; proofAt: number; nonce: string }>;
};

describe("transport request proof vectors (owner-signed)", () => {
	const { ownerSignPriv } = vectors;

	it("reproduces the canonical TRANSPORT_REQUEST_V1 bytes + signature (cross-runtime pin)", () => {
		const { signerSignPub, proofAt, nonce } = vectors.transport.value;
		const bytes = transportRequestSigningBytes(signerSignPub, proofAt, nonce);
		assertCanonicalBytes(bytes, vectors.transport);
		// Deterministic Ed25519: re-signing with the fixed key reproduces the pinned signature.
		const signed = signTransportRequest(signerSignPub, proofAt, nonce, ownerSignPriv);
		expect(signed.proof).toBe(vectors.transport.signature);
		expect(verifyTransportRequest(signed)).toBe(true);
		expect(verify(bytes, vectors.transport.signature, signerSignPub)).toBe(true);
	});

	it("parses as a TransportRequest and rejects a tampered proof under the wrong signer", () => {
		const { signerSignPub, proofAt, nonce } = vectors.transport.value;
		const signed = signTransportRequest(signerSignPub, proofAt, nonce, ownerSignPriv);
		expect(TransportRequestSchema.safeParse(signed).success).toBe(true);
		// A captured proof re-pointed at a different claimed signer no longer verifies.
		const forged = { ...signed, signerSignPub: "AAAA" };
		expect(verifyTransportRequest(forged)).toBe(false);
	});

	it("the distinct version tag stops a roster proof from verifying as a transport request", () => {
		// Same key + proofAt + nonce, but ROSTER_V1 vs TRANSPORT_REQUEST_V1, so neither proof crosses over.
		const { signerSignPub, proofAt, nonce } = vectors.transport.value;
		const rosterSig = signRosterRequest(signerSignPub, proofAt, nonce, ownerSignPriv);
		expect(verify(transportRequestSigningBytes(signerSignPub, proofAt, nonce), rosterSig, signerSignPub)).toBe(
			false,
		);
	});
});

File: android/app/src/test/java/com/atelier_nyaarium/switchboard/crypto/TransportRequestTest.kt
package com.atelier_nyaarium.switchboard.crypto

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Pins the Kotlin transport-request signing against switchboard's node:crypto, through the same
 * vectors the vitest suite reads (tests/fixtures/transport-request/vectors.json). If the canonical
 * encoding or the signature scheme diverges, the node-signed proof below stops verifying here -
 * which would silently break an owner pulling its network's gateway-bridge transport. Mirrors
 * ProvisionOpsTest's roster / trust-pending vectors.
 */
class TransportRequestTest {
	private val json = Json { ignoreUnknownKeys = true }

	private fun vectors() =
		json.parseToJsonElement(
			javaClass.classLoader!!.getResourceAsStream("transport-request/vectors.json")!!.bufferedReader().readText(),
		).jsonObject

	@Test
	fun transportRequestCanonicalBytesMatchNode() {
		val v = vectors()
		val vec = v["transport"]!!.jsonObject
		val value = vec["value"]!!.jsonObject
		val signer = value["signerSignPub"]!!.jsonPrimitive.content
		val bytes = ProvisionOpsCrypto.transportRequestSigningBytes(
			signer,
			value["proofAt"]!!.jsonPrimitive.content.toLong(),
			value["nonce"]!!.jsonPrimitive.content,
		)
		CanonicalBytes.assertCanonicalBytes(bytes, vec)
		// The node-signed proof verifies under the signer key (the twin reproduces it).
		assertTrue(Crypto.verify(bytes, vec["signature"]!!.jsonPrimitive.content, signer))
		// A different key must not verify the node-signed proof.
		assertFalse(Crypto.verify(bytes, vec["signature"]!!.jsonPrimitive.content, Crypto.generateIdentity().sign.pub))
		// The distinct version tag means the ROSTER_V1 proof never verifies as a transport request.
		val rosterSig = ProvisionOpsCrypto.signRosterRequest(
			signer,
			value["proofAt"]!!.jsonPrimitive.content.toLong(),
			value["nonce"]!!.jsonPrimitive.content,
			v["ownerSignPriv"]!!.jsonPrimitive.content,
		)
		assertFalse(Crypto.verify(bytes, rosterSig, signer))
	}

	@Test
	fun signsAndVerifiesLocally() {
		val owner = Crypto.generateIdentity()
		val attacker = Crypto.generateIdentity()
		val proof = ProvisionOpsCrypto.signTransportRequest(owner.sign.pub, 9000L, "bm9uY2U=", owner.sign.priv)
		val bytes = ProvisionOpsCrypto.transportRequestSigningBytes(owner.sign.pub, 9000L, "bm9uY2U=")
		assertTrue(Crypto.verify(bytes, proof, owner.sign.pub))
		assertFalse(Crypto.verify(bytes, proof, attacker.sign.pub))
	}
}
